경북대 FE 장세현 1단계 - 선물하기 메인 API 구현하기

SeHIgh opened the pull request:
안녕하십니까 경북대학교 프론트엔드 과정 수강중인 장세현입니다.
1단계 과제 - 선물하기 메인 API 구현하기 내용을 PR 요청 드립니다!

---

### 과제 진행 사항

- 선물하기 홈 > 선물 테마 섹션
  - /api/themes API를 사용하여 테마 목록을 완성
  - 데이터를 불러오는 동안 로딩 화면 구현
  - 데이터가 없거나, 에러가 발생하면 선물 테마 섹션이 보여지지 않게 하기

- 선물하기 홈 > 실시간 급상승 선물랭킹 섹션
  - /api/products/ranking API를 사용하여 실시간 급상승 선물 랭킹 섹션 구현
  - API 명세를 살펴보고 각 필터 선택 시 해당 필터에 맞는 API를 재 요청 하기
  - 데이터를 불러오는 동안 로딩 화면 구현
  - 보여 줄 상품 목록이 없을경우 상품 목록이 없다는 문구를 보여주기

- 본인만의 기준으로 일관된 코드 작성
- 기능 단위로 나누어 커밋
- 재사용을 고려하여 API 구현

SeHIgh committed:
docs(README, axios): 1단계 과제 요구사항 정리 및 axios 설치

SeHIgh committed:
feat: API 구조 개선 및 카테고리 기능 리팩터링

- api/services, api/types 디렉토리 및 관련 파일 추가로 API 폴더 구조 개선
- 공통 데이터 fetch를 위한 useFetch 커스텀 훅 구현
- 카테고리 컴포넌트가 목 데이터 대신 API 데이터를 사용하도록 변경

SeHIgh committed:
feat(category): 데이터가 없거나, 에러가 발생하면 선물 테마 섹션이 보여지지 않도록 구현

SeHIgh committed:
feat(loading): 로딩 컴포넌트 추가 및 카테고리 컴포넌트에 로딩 상태 처리 구현

SeHIgh committed:
feat(trend): 상품 랭킹 fetch 서비스 및 UI 개선

- 상품 랭킹 데이터를 불러오는 fetch 서비스 구현 (api/services, api/types 등)
- 트렌드(랭킹) 섹션에서 API 데이터를 활용하도록 개선
- 로딩 중에는 로딩 컴포넌트가 표시되도록 처리
- 빈 목록(데이터 없음) 상황에 대한 예외 처리 및 안내 메시지 구현

Leejin-Yang commented on src/api/services/product.ts:
```diff
@@ -0,0 +1,17 @@
+import apiClient from '@/api'
+// import { RankType, TargetType } from '@/api/types/product'
```
주석처리가 남아있네요.!

Leejin-Yang commented on src/api/services/product.ts:
```diff
@@ -0,0 +1,17 @@
+  targetType: TargetType = 'ALL',
+  rankType: RankType = 'MANY_WISH',
+): Promise<Product[]> => {
+  const res = await apiClient.get<{ data: Product[] }>('/api/products/ranking', {
```
api path들도 상수로 관리하면 더 좋아보입니다 😊

Leejin-Yang commented on src/features/product/components/trend/Trend.tsx:
```diff
@@ -29,9 +30,22 @@ export const Trend = () => {
+    isLoading,
+    isError,
+    data: products,
+  } = useFetch<Product[]>(() => fetchProductRankList(targetType, rankType))
```
[중요] 익명 화살표 함수가 새로운 참조로 넘어가서 무한으로 요청을 보내고 있어요. 메모이제이션을 통해 각 파라미터가 바뀔때 함수를 업데이트 하도록 하면 어떨까요?

Leejin-Yang commented on src/features/product/components/trend/Trend.tsx:
```diff
@@ -29,9 +30,22 @@ export const Trend = () => {
+  } = useFetch<Product[]>(() => fetchProductRankList(targetType, rankType))
+
   // * 초기 보여줄 상품 개수
   const INITIAL_SHOW_COUNT = 6
```
이 값은 컴포넌트 외부에 작성해도 좋아보이네요

Leejin-Yang commented on src/features/product/components/trend/Trend.tsx:
```diff
@@ -81,11 +84,52 @@ export const Trend = () => {
+  // * 빈 목록 or 에러 화면
+  if (isError || !products || products.length === 0)
+    return (
+      <Container>
+        {/* 실시간 급상승 필터 컨테이너 */}
+        <h1 css={theme.typography.title.title1Bold}>실시간 급상승 선물랭킹</h1>
+
+        {/* 필터 */}
+        <TrendFilter
+          targetType={targetType}
+          rankType={rankType}
+          onTargetTypeChange={handleTargetTypeChange}
+          onRankTypeChange={handleRankTypeChange}
+        />
```
로딩, 에러 상태, 성공 상태에 따라 렌더링 되는 ui에 공통되는 부분을 중복으로 작성하고 있는데요, 조건부 렌더링 영역을 줄여보는건 어떨까요?

SeHIgh replied:
저도 이 중복되는 부분 때문에 고민을 하다가 컴포넌트를 또 분리하면 배보다 배꼽이 커지는 느낌이라 그냥 중복된 코드를 사용하는 방향으로 작성하였습니다.

그래서 이번에 멘토님의 피드백을 받고 다시 고민하다가
구글 서치와 llm의 힘을 빌려서
아래와 같은 형태로 개선 해보았습니다!

```
// * 바뀌는 영역 조건부 관리
  let body: React.ReactNode
  if (isLoading) {
    // * 로딩 화면
    body = (
      <LoadingContainer>
        <Loading />
      </LoadingContainer>
    )
  } else if (isError || !products || products.length === 0) {
    // * 빈 목록 or 에러 화면
    body = (
      <LoadingContainer>
        <EmptyMsg variant="subtitle2Regular">상품 목록이 없습니다.</EmptyMsg>
      </LoadingContainer>
    )
  } else {
    body = (
      <>
        <ProductContainer>
          {displayProducts.map((product, index) => (
            <ProductItem key={product.id} product={product} index={index} />
          ))}
        </ProductContainer>

        {/* 더보기 버튼 */}
        {shouldShowMoreButton && (
          <MoreButtonContainer>
            <Button variant="outline" size="medium" onClick={handleMoreButtonClick}>
              {showAll ? '접기' : `더보기`}
            </Button>
          </MoreButtonContainer>
        )}
      </>
    )
  }

  return (
    <Container>
      {/* 실시간 급상승 필터 컨테이너 */}
      <h1 css={theme.typography.title.title1Bold}>실시간 급상승 선물랭킹</h1>

      {/* 필터 */}
      <TrendFilter
        targetType={targetType}
        rankType={rankType}
        onTargetTypeChange={handleTargetTypeChange}
        onRankTypeChange={handleRankTypeChange}
      />

      {/* 조건부 렌더링 */}
      {body}
    </Container>
  )
```

이렇게 하니 별도 컴포넌트 생성 없이도 간단하게 조건부 렌더링을 구현할 수 있어서 좋은 것 같습니다!

Leejin-Yang replied:
관련해서는 이 코멘트 확인해주세요 😊

https://github.com/next-step/react-gift-product-list/pull/65#discussion_r2210715785

Leejin-Yang commented on src/components/category/Category.tsx:
```diff
@@ -1,20 +1,38 @@
+  if (isLoading)
+    return (
+      // 외부 컨테이너
+      <Container>
+        {/* 카테고리 타이틀 */}
+        <h1 css={theme.typography.title.title1Bold}>선물 테마</h1>
+        {/* 로딩 서브 컨테이너 */}
+        <LoadingSubContainer>
+          <Loading />
+        </LoadingSubContainer>
+      </Container>
+    )
```
카테고리에서도 적용해 볼 수 있겠네요 😊

Leejin-Yang commented on src/api/services/product.ts:
```diff
@@ -0,0 +1,17 @@
+
+// * 상품 랭킹 목록 조회하기
+export const fetchProductRankList = async (
+  targetType: TargetType = 'ALL',
+  rankType: RankType = 'MANY_WISH',
```
작성한 상수에서 값을 가져와보면 어떨까요?

Leejin-Yang commented on src/api/services/product.ts:
```diff
@@ -0,0 +1,17 @@
+export const fetchProductRankList = async (
+  targetType: TargetType = 'ALL',
+  rankType: RankType = 'MANY_WISH',
+): Promise<Product[]> => {
```
Product 도 외부 데이터이므로 api 폴더에서 타입을 관리하면 좋아보이는데 어떻게 생각하시나요?

SeHIgh replied:
기존에는 목데이터를 사용하면서 api 폴더가 따로 없어서 feature에서 전체 관리를 하였었는데, 이번에 api 요청을 사용하게 되면서 별도로 타입을 만들까 하다가 기존에 동일한 내용으로 작성이 되어있었기 때문에 그대로 두었었는데,
멘토님 말씀대로 api 폴더에서 관리하도록 하는게 좋을 것 같다 생각됩니다!

Leejin-Yang commented on src/api/index.ts:
```diff
@@ -0,0 +1,12 @@
+import axios from 'axios'
+
+const BASE_URL = import.meta.env.VITE_API_BASE_URL
```
👍

Leejin-Yang commented on src/components/ui/Loading.tsx:
```diff
@@ -0,0 +1,35 @@
+import { LoaderCircle } from 'lucide-react'
+
+// * 로딩 스피너 컴포넌트
+export const Loading = () => (
```
로딩 컴포넌트도 잘 만들어주셨네요 👍

Leejin-Yang reviewed (changes requested):
세현님 안녕하세요 👋 이번 단계 잘 진행해주셨네요
fetch 함수들도 깔끔하게 잘 정리해주셨다는 생각이 듭니다

다만, 요청 보내는 과정에 대해 코멘트 남겼는데 이 부분 확인해주세요
다른 코멘트들도 참고해주시고 다시 요청 보내주세요 😊

SeHIgh committed:
feat(product): 타입 정의 및 유효성 검증 함수 추가, 기존 타입 제거

- product 관련 타입 파일 위치 이동 (api)
- 기존 불필요한 형태의 타입을 enum 형태로 개선 & 그에 맞게 유효성 검증 함수도 수정
- 상수 형태로 접근이 가능해진 점을 이용하여 각종 하드코딩된 부분을 상수로 개선

SeHIgh committed:
feat(api): API 엔드포인트 상수화 및 상품/테마 조회 함수 적용

SeHIgh committed:
feat(product): UI 관련 상수 파일 분리 및 적용

SeHIgh committed:
fix(types): 빌드 시 감지된 오류 수정 (Product import 경로 수정)

SeHIgh committed:
refactor(trend): 조건부 렌더링 영역 관리를 통해 코드 가독성 향상

SeHIgh committed:
refactor(category): 조건부 렌더링 영역 관리를 통해 코드 가독성 향상

SeHIgh committed:
fix(fetch): 의존성 배열 추가로 데이터 패칭 최적화

- useFetch 커스텀 훅에 의존성 배열 기반 메모이제이션 추가

SeHIgh commented:
멘토님 꼼꼼한 피드백 정말 감사합니다!
멘토님의 피드백을 반영하여 코드들을 수정해보았습니다!

### 수정 사항

1. enum 기반 타입 정의
  - 문자열 리터럴 타입을 enum으로 변경하여 타입안정성을 강화하고 사용성을 개선하였습니다!
  - 추가로 해당 enum 기반 타입을 이용하여 별도 파일에서 상수로 사용할 수 있도록 하였습니다!
  - 추가로 유효성 검증 함수를 enum 기반 타입 검증 형태로 개선하였습니다.

2. 타입 파일 구조 개선
  - `features/product/types` → `api/types/product`로 이동하여 API 관련 타입을 통합 관리할 수 있도록 개선하였습니다.

3. 상수 관리 체계화
  - API 엔드포인트 상수화 하여 하드코딩된 API 경로를 `api/constants/endpoints.ts`로 관리하도록 개선하였습니다.
  - UI 상수도 분리하여 컴포넌트 내 하드코딩된 값을 `features/product/constants/ui.ts`로 분리하여 관리하도록 하였습니다.

4. 성능 최적화
  - `useFetch` 훅에 의존성 배열 기반 메모이제이션을 추가하여 데이터 패칭 최적화를 적용하였습니다!
  -  패칭을 사용하는 컴포넌트 내에서 `useCallback`을 사용하는 방법도 있겠지만, 근본적인 해결을 통해 추후 더 깔끔하고 간편한 코드를 위해 `useFetch` 훅을 개선하였습니다.

6. 코드 가독성 향상
  - 중복된 JSX 구조를 통합하여 조건부 렌더링을 통해 유지보수성을 향상 시켰습니다.

Leejin-Yang commented on src/api/constants/endpoints.ts:
```diff
@@ -0,0 +1,12 @@
+// * API 엔드포인트 통합 상수 관리
+export const API_ENDPOINTS = {
```
👍

Leejin-Yang commented on src/api/types/product.ts:
```diff
@@ -0,0 +1,45 @@
+// * 타겟 타입
+export enum TargetType {
+  ALL = 'ALL',
+  FEMALE = 'FEMALE',
+  MALE = 'MALE',
+  TEEN = 'TEEN',
+}
+// * 랭크 타입
+export enum RankType {
+  MANY_WISH = 'MANY_WISH',
+  MANY_RECEIVE = 'MANY_RECEIVE',
+  MANY_WISH_RECEIVE = 'MANY_WISH_RECEIVE',
+}
```
enum이 어떤 장단점이 있는지도 학습해보면 좋아보여요 😊

Leejin-Yang commented on src/features/product/components/trend/Trend.tsx:
```diff
@@ -81,9 +90,43 @@ export const Trend = () => {
-    // TODO: 추후 실제 상품 목록 필터링 시 수정
-    console.log('Selected rank type:', type)
+  // * 바뀌는 영역 조건부 관리
+  let body: React.ReactNode
+  if (isLoading) {
+    // * 로딩 화면
+    body = (
```
fetch 상태에 따라 잘 분리해주셨네요 👍
다만, 컴포넌트 내부에 로직이 많아질수록 가독성을 신경써주어야 할 수 있어요. 이 부분을 선언적으로 할 수 있는 방법도 고려해보면 좋아보여요 😊

예를 들어 렌더하는 함수로 분리하거나, 객체를 활용해볼 수 있겠네요

Leejin-Yang commented on src/features/product/constants/ui.ts:
```diff
@@ -0,0 +1,4 @@
+// * 상품 UI 관련 상수
+export const PRODUCT_UI_CONSTANTS = {
+  INITIAL_SHOW_COUNT: 6,
+} as const
```
잘 분리해주셨네요 👍 
만약 사용처가 한 곳이라면 컴포넌트 선언전에 작성해주어도 좋아보여요

Leejin-Yang commented on src/features/product/components/trend/Trend.tsx:
```diff
@@ -17,20 +24,33 @@ export const Trend = () => {
+
   // * 초기 보여줄 상품 개수
-  const INITIAL_SHOW_COUNT = 6
+  const INITIAL_SHOW_COUNT = PRODUCT_UI_CONSTANTS.INITIAL_SHOW_COUNT
```
그렇다면 이 부분이 Trend 컴포넌트 선언 윗 라인으로 갈 수 있겠네요 😊

Leejin-Yang reviewed (approved):
피드백 잘 반영해주셨네요 👍
요구사항도 모두 만족해주시고 구조적으로도 잘 작성해주신걸로 보여서, 이번 단계는 여기서 마무리해도 좋아보여요
몇가지 코멘트 남겨두었는데, 반영할 부분은 다음 단계 시작하면서 반영해주세요
다음 단계에서 보아요 🙌
